### PR TITLE
[3.0-dev] Major refactor of serialization in prep for 3.0

### PR DESCRIPTION
**Pull request by bchess**
(note this PR is just for merging into the `v3.0.0-dev` branch, not main)

# file schema
The file schema has changed. Immediately following the metadata block is the full block of all header structures.  

After the metadata and headers is the block of raw tensor data. The block of tensor data begins on a 4k page-aligned boundary (relative to start of file). Each individual tensor begins on an 8-byte-aligned boundary.

Basically:
```
file_header
metadata_size
metadata0
...
metadataN
header0
...
headerN
tensor0 (file_offset % 4096 == 0)
tensor1 (file_offset % 8 == 0)
...
tensorN (file_offset % 8 == 0)
```
We may want to combine the metadata and the header entries, but they are left separate for now for easier backwards compatibility.

## hash computation
Previously, header hash computations included the hash fields themselves, presumed to be zeroed. This has been changed so the hash computation segments omit the hash fields.  ~This breaks `verify_hashes=True` with 2.x files for now. We should prob fix prior to release~ (This is now fixed, we can deser 2.x files with verify_hashes)

# bulk_write refactor
`_bulk_write()` has been broken up into a series of smaller functions that each perform some operation on the tensors. `_WriteSpec` carries the state of each tensor as it gets processed. Many ops are threaded and tracked as Futures (`_WriteSpec.tensor_data_task`). Dependent steps do Future chaining.

## tensor_write()
`tensor_write()`, e.g. "incremental mode", is still supported. (It is implemented as a call to `_bulk_write([tensor])`). There are new tests to validate its functionality.

### max_tensors
Because the new file schema must know ahead of time where to start writing tensor data, there must be a certain amount of pre-allocated space for the metadata and headers. Thus you must specify `max_tensors` to `TensorSerializer.__init__` If you plan on callling `tensor_write()` more than once. This does a wild-ass-guess estimate on a good amount of header space to reserve. In some cases, like really long tensor names, it could still be an underestimate. You will get a RuntimeError() if you call `write_tensor()` and there isn't enough pre-allocated space.

You do not need to set `max_tensors` if you just call `tensor_write()` once.

# Performance
Note that while these changes are mainly motivated to improve performance, I did not spend much time in this PR to measure and optimize performance. There may be regressions, but we should be able to resolve them (and beyond!) before release.

### CODE REVIEW

File: tensorizer/serialization.py
Comment (harubaru) [inline tensorizer/serialization.py:3785]: Where did this assertion go?

File: tensorizer/serialization.py
Comment (harubaru) [inline tensorizer/serialization.py:4225]: What are the consequences of removing this argument? Do we know if anyone is using this?

File: tensorizer/serialization.py
Comment (harubaru) [inline tensorizer/serialization.py:2930]: Why is this a `KeyError` instead of a `ValueError`?

File: tensorizer/serialization.py
Comment (harubaru) [inline tensorizer/serialization.py:3336]: Can there be added docs on this?

**Comment by harubaru** [comment]
## Feedback

The code looks pretty good and it also works pretty good too. I have tested this against various serialization tasks by serializing models for inference which did not result in a failure. The comments that I have left on this are mainly to get a better understanding behind the intent of some of the changes in this PR even if it was as little as changing an error return value.

That being said, the changes to the file schema that this PR provides are nice such as aligning tensor data on fixed boundaries or the decomposition of `bulk_write()`. Though, I am a bit curious about the addition of `max_tensors`. Is `max_tensors` a necessary argument to have because we support `write_tensor()`?

Additionally, it would be helpful to understand `max_tensors` more with a code sample that demonstrates its necessity, just like the code samples in the other docstrings.

## Testing

A good portion of the changes in this PR and the PR's intent is changing the file schema. I have spent a good portion of my time reviewing this PR by testing the changes that were implement. The PR works with deserializing older 2.x artifacts and I have gathered benchmarking data from serializing and deserializing GPT-J 6B for 50 trials which did not show a significant performance difference between this PR and 2.9.0 (or at least: I don't think it is a significant difference.)

image

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:4285]: I would choose the non-cloned tensor to be the last one, because its `tensor_data_task` will block the longest, so that would unblock some processing for a subsequent stage sooner, considering the access pattern of `write_specs` is consistently first-to-last.
```suggestion
            for w in shared_write_specs[:-1]:
                clone_tasks.append(
                    self._computation_pool.submit(
                        self._do_clone, w, clone_dependencies
                    )
                )
                w.user_owns_tensor_data = False

            shared_write_specs[-1].tensor_data_task = _FutureGroup(
                clone_tasks + clone_dependencies.futures
            )
            for w in shared_write_specs[:-1]:
                w.tensor_data_task = _FutureGroup(clone_tasks)
```

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:4279]: Adding a clone task can be skipped if a tensor is being cloned for not having been contiguous. Since that happens asynchronously, it should still be included up to this point (so its cloning status is tracked by `clone_dependencies`); it just won't need the final cloning part here.

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:4258]: Interestingly, PyTorch counts tensors imported from other libraries like `numpy` as always having separate storages, even if they overlap or represent a completely identical region when compared with other storages—even in its `Tensor.is_set_to` function, so non-PyTorch-native tensors are sort of a "known bug" for this method, though it's niche enough to not affect almost any normal workflows.

Other than that, we can identify overlapping input data more precisely using an $O \lparen n \cdot \log \lparen n \rparen \rparen$ (where $n$ is the number of tensors) algorithm I designed for it once, since this old way incorrectly identifies disjoint slices of a tensor as being shared, but we can add that in another PR, since it's not a regression of this one.

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:4248]: I'd try to avoid `assert` statements for anything regarding encryption correctness (or anything else important) since `assert` statements can be skipped when optimization is enabled, just in case.
```suggestion
        if not self._encrypted or self._encryption is None:
            raise RuntimeError(
                "Tried to encrypt tensors without encryption parameters"
                " having been provided"
            )
```

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:4298]: This should not need to block the main thread on `tensor_data_task`. Do you need me to change the `_crypt.ChunkedEncryption` interface to make this possible? I can probably make an option to provide the buffer after constructing it, as long as the length is provided in advance.

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:4347]: The comment `# bogus file_offset. This gets filled in in build()` isn't accurate; it just gets filled in a bit later in this function but before `build()`. It's the tensor data offset that gets filled in in `build()`.
```suggestion
            w.header = _TensorHeaderSerializer(
                w.module_index,
                w.tensor_type,
                w.name.serialized_(),  # name as bytes
                dtype_bytes,
                w.shape,
                w.data_length,
                0,  # placeholder file_offset
                include_crc32=w.include_crc32,
                include_sha256=w.include_sha256,
                crypt_info=w.crypt_info,
            )
```

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:1]: This is more of a general comment as it doesn't block this PR, but as an optimization we should operate on detached copies of tensors to better control tensor lifetime and their garbage collection impact (by manually `set_`-ing them to empty buffers at the end of the serialization process). The handling of futures in this PR makes the object lifetime a little unclear, but depending on what function deallocation triggers in, it can reduce performance by a bit. Plus, deallocation can even be shunted to a background thread right before returning from the function that way.

Comment (Eta0) [inline tensorizer/serialization.py:1]: This one isn't a blocker for merging this PR, but an extra thought (doesn't require an answer either): Does queueing operations in batches to repeatedly loop over the entire list of tensors rather than perform many different operations in order on a each tensor negatively impact performance due to the worse memory access pattern? For small data it would, but I don't know if tensors are so large that memory caching wouldn't affect them much anyway. Hashing performance, for example, tends to improve when a tensor is split into chunks and alternating hashes are performed on the chunks, so I'm wondering how significant it might be to be able to fit operations like "hash a block of data, encrypt it, and write it" as back-to-back steps at some point. Defining compound operations like `stream_decrypt` was helpful in the deserializer, so something like that might be able to work in this organizational scheme too.

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:4224]: Does this make sense to include if all new serialized files are past this version anyway? Or are we planning on supporting serializing with older formats and potentially sharing this code path?

File: tensorizer/serialization.py
Comment (Eta0) [inline tensorizer/serialization.py:1734]: We should probably read the entire metadata section including these headers in one or two `read` (or `readv`) operations and just use the `_TensorHeaderDeserializer` buffer constructor instead of `from_io`.

File: tests/test_serialization.py
Comment (Eta0) [inline tests/test_serialization.py:815]: It used to be the case that a bulk write could still be incremental, and a series of bulk writes could be used to create a tensorizer file. Bulk write functions are also the only way to incrementally write structured data, namely through `write_state_dict`. As far as I can tell, the way the serialization code is written still allows that, so we should add some test cases for that here too.

**Review by Eta0 — CHANGES_REQUESTED** [review]
# Design Considerations
Separate from the issues in the comments, there are three points that require the most significant changes or design consideration to accompany this. Most of these are not "these need to be implemented in this PR" but "the designs we will need for these should be considered to see if the changes we are making now will be compatible with them," and is more for discussion.

## Accounting for Header Encryption
Tensorizer v3 should extend encryption to include header information. This needs to be accounted for in the file metadata and the patterns we use to buffer and write the metadata.
Important things to consider:
1. Where do we put encryption parameters in the header?
	- My take: Indicate in the file flags if the headers are encrypted, and then place encryption info immediately before the metadata section if so. It likely shouldn't be bundled into the metadata section, so that the length of the metadata section can also be encrypted. File flags and other meta-metadata like hashes of the metadata section can be encrypted too.
2. How does incremental writing interact with metadata encryption?
	- How many nonces and MACs are used when encrypting and decrypting the metadata section?
	- If one is used, then the metadata section can only be updated by rewriting the entire thing, because data can't be inserted into the middle of an encrypted stream (for the non-tensor-header metadata section), only appended at the end.
		- This might be cheap enough that we could do it anyway, though it feels like it could be limiting in the future.
	- If two are used (one for metadata entries, one for tensor headers), the metadata section could actually be written as two independent streams, only rewriting their MACs on each synchronization.
	- If multiple are used, then we need to handle size limits for the nonce & MAC list like size limits of the rest of the metadata section. (Luckily, this fits pretty well, since the size taken up by nonces and MACs is directly proportional to the number of entries being described when using one encrypted segment per metadata entry).
	- Ideally, information like the choice of encryption method would be stored in a tagged and extensible format like tensor `CryptInfo` segments, which would allow changing this in order to make it more or less rigid in the future.
3. To what extent do we attempt to protect information about model "size"?
	- The filesize gives away most information about that regardless of what we encrypt, but should we set out to protect information about how many tensors are contained in a file?
	- The number of tensors could potentially be leaked by information like the length of the metadata section, or the spacing of padding between tensor data entries. The length of the metadata section can be determined by the padding at the end of the metadata section unless we scramble the padding bytes during encryption.

We don't need to have functional header encryption to merge this PR, but deciding how it will fit into the metadata will help this one not conflict with that change when we do it later. I can work on the implementation for it later once it's more figured out.

## Accounting for File Range Insertion
A few serialization media allow inserting more space into the middle of a file instead of just at the end. In particular, `ext4` allows it, which accounts for much of local file serialization, and serialization to temporary files on a local disk, which accounts for object storage and (potentially) Azure blobstore serialization. If we eventually support object storage uploads without temporary files, those would similarly not limit adding more data into the middle of a file. For these cases we can do any header modifications we want without worrying about determining the required metadata size before writing tensor data, or we could even shrink over-allocated space in the metadata section to match what was written (deleting a file range).

For concerns about supporting this:
- Most filesystems only support adding new data ranges at logical block boundaries, so the metadata entry and header entry sections need to be separated by a page boundary when doing this.
	- They don't need to be separated that far in general, so this can be decided on-the-fly as part of the implementation, which ought to be compatible with this design, to my understanding.
- We could write data offsets into headers relative to the beginning of the data section of the file, so that we don't need to know the length of the metadata section in order to write headers. The beginning of the data section of the file ought to be known to the deserializer at the point of reading those anyway, so it could add that to each offset as soon as each header is deserialized, and there wouldn't need to be a significant amount of code changed in the deserializer.

## More Granular Future Dependencies
In this PR's changes, dependencies for later tasks are tracked as a single `tensor_data_task`. We should add differentiation for write hazards and read hazards, since storing it as one item doesn't allow two operations that only need to read tensor data to happen simultaneously (e.g. computing hashes while also writing data to disk, when encryption isn't active). This is less harmful to performance when queueing operations in batch order like now rather than per-tensor, but any time that there are enough threads (or few enough tensors) to handle tasks from multiple stages at once, this would unblock the later stages sooner.

In addition to those three things, there was also this optional suggestion, which would complement the first two somewhat:
## Optional Syncing
We could add a `synchronize=False` mode (referring to syncing data to disk) for incremental writes that defers writing headers until a write occurs with `synchronize=True` or the file is finalized. For cases like uploading to a temporary file before uploading to S3, writing headers can always be deferred, and this could be better for performance whenever constant synchronization isn't needed since it can guarantee a single write for that entire section, and (potentially) allow less fiddling with encryption information in the header section from having to repeatedly edit any data in it (like MACs).

## Future Work
Finally, these are various things on my v3 wishlist that aren't directly relevant to this PR and aren't yet implemented (nor should they be addressed here), but I wanted to make a note of them here for future reference.
- Add cryptographic authentication / checksums / ECC (if we're getting fancy) for the entire metadata and header section of the file, as a unit, since this allows checking that the metadata is complete and not missing entries (which could potentially happen if, say, the metadata length segment wasn't set properly at serialization time, or ended up changing somehow).
- Use more consistent and documented exception types
	- E.g. some sort of `InvalidTensorizerFile` exception when deserializing, which could have subtypes like `UnsupportedTensorizerVersion`
	- This would just make it a bit easier for code using tensorizer to differentiate between bad files or issues with their system, library dependency versions, deserialization flags they requested, or various other things
	- Having a smaller, documented set of errors in the outer layer of the API makes it so people don't have to use `except Exception` as often, which is usually nice
- Remove the `compress_tensors` option to the serializer and its vestigial related machinery, since it has done nothing even as far back as tensorizer v1.0.0
- Turn `TensorSerializer` into a context manager to match with `TensorDeserializer`
- I need to tweak `ChunkedEncryption` slightly in a way that is incompatible with older deserializers, so I'd like to add it in the new data version 5 alongside the changes in this PR

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:537]: imo metadata is a dead-end. Keeping it intact for now makes it easier to load 2.x tensors, but in future major versions everything would be in header. Organizationally they are already the same as headers in 3.x

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:4420]: Agreed!

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:3821]: 💵 💵

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:1]: I don't know, but my intuition is that it's minimal impact. Looping over tensors in aggregate makes it easier to handle batched operations or relationships/dependencies between tensors though.

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:4224]: Yeah its fine to remove

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:4225]: I don't really know why this was useful imo, but open to pushback. In a major version update we can break the interface

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:3336]: Yeah agreed there should be some docs on this

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:2930]: Updated!

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:3336]: Right now refer to my info in the PR description itself, but we should include something like that in the code

**Comment by bchess** [comment]
> ## Accounting for File Range Insertion

Imo its a fair bit of added code complexity, adds more constraints to what we can optimize, for little benefit. Object stores upload APIs have minimums to the chunk sizes, and you still need to assign the correct index of the block ahead of time -- you can't insert a block later.

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:3870]: Yeah, but the logic inside `_maybe_fallocate()` uses its `tensors` list to calculate total size. To work in incremental mode it'd need to be updated to account for past writes.

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:3904]: This is a good point. +1 to adding unit tests, because feels like something that could easily regress

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:4298]: The dependency itself is for `tensor_memory.nbytes` which then informs chunk count. I think you're right that the whole block could be pulled out into a thread, but it gets a little ugly that we'd start setting other attributes (`crypt_info` and `encryptor`) of `WriteSpec` on a thread without a more sophisticated mutex design. Correct me if I'm wrong, but I don't think there's anything here that's CPU intensive and going to be able to release the GIL. And we'd need all to reconcile by the next step of `_prepare_for_write_headers` anyway.

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:1734]: Agreed!

**Comment by bchess** [comment]
In terms of dev process:

There are a lot of good suggestions on changes and improvements, but this PR is already uncomfortably big. I think it's easier to make those changes as subsequent and individual PRs against the 3.x branch. I can take point on writing up issues on all of the pending comments so that they don't get lost. For some of the spicier ones, we can also use the issues as separate threads of discussion on each topic. How does that sound?

File: tensorizer/_futuregroup.py
Comment (bchess) [reply tensorizer/_futuregroup.py:25]: https://github.com/coreweave/tensorizer/issues/141

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:4437]: https://github.com/coreweave/tensorizer/issues/143

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:535]: https://github.com/coreweave/tensorizer/issues/142

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:3916]: https://github.com/coreweave/tensorizer/issues/144

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py]: https://github.com/coreweave/tensorizer/issues/145

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:4275]: https://github.com/coreweave/tensorizer/issues/146

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py]: https://github.com/coreweave/tensorizer/issues/148

File: tensorizer/serialization.py
Comment (Eta0) [reply tensorizer/serialization.py:4225]: Once upon a time, this could potentially progressively deallocate memory by consuming a module, but it was sort of at the whims of the garbage collector, and wouldn't really reduce peak RAM usage anyway, so it wasn't very useful. This used to be the default behaviour, and the first time I started engaging with the tensorizer repo was to help someone debug this behaviour, because it was unintuitive and messing their stuff up. I don't see too much value to keeping it, especially since parallelized bulk writing tends to keep references to tensors around for much longer.

If anyone does need progressive deallocation of memory, this would probably be better implemented as a generalized callback hook for when each tensor has finished being processed, which would let other code deallocate tensor memory *and actually know that it has happened so that they can start taking advantage of it*... among other use cases. Since this is a blocking interface, it's currently unlikely that anyone could use it more effectively than they could write `del m` after the call returns, so the workaround for any code currently using it just to free memory at some nonspecific point shouldn't be terribly inconvenient.

File: tests/test_serialization.py
Comment (bchess) [reply tests/test_serialization.py:815]: https://github.com/coreweave/tensorizer/issues/149

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:3882]: https://github.com/coreweave/tensorizer/issues/150

File: tensorizer/serialization.py
Comment (Eta0) [reply tensorizer/serialization.py:537]: That's fair. The fixed format for metadata has been problematic before because e.g. `data_length` would be far more useful if split into `serialized_length` and `deserialized_length` to separately communicate the file layout and the amount of RAM/VRAM usage the deserializer will need to anticipate. If it is only used as a shortcut to navigate the more detailed header section that follows it, then I'd be equally fine with it being a mostly-dead section (or just removing it now from 3.0 files and keeping only its deserializer).

File: tensorizer/serialization.py
Comment (Eta0) [reply tensorizer/serialization.py:3882]: `fallocate` extends the file relative to the current file position, so it doesn't need anything special to account for past writes if the file position is reliable; it can just progressively extend the file.

**Comment by bchess** [comment]
> In terms of dev process:
> 
> There are a lot of good suggestions on changes and improvements, but this PR is already uncomfortably big. I think it's easier to make those changes as subsequent and individual PRs against the 3.x branch. I'd also like to empower others to be able to make contributions as well 😜
> 
> I can take point on writing up issues on all of the pending comments so that they don't get lost. For some of the spicier ones, we can also use the issues as separate threads of discussion on each topic. How does that sound?



> > ## Accounting for File Range Insertion
> 
> ~Imo its a fair bit of added code complexity, adds more constraints to what we can optimize, for little benefit. Object stores upload APIs have minimums to the chunk sizes, and you still need to assign the correct index of the block ahead of time -- you can't insert a block later.~
> 
> Thinking about this more, I maybe have presumed different from what you were proposing. I don't think we should add a public interface that allows for the mid-range insertion of a tensor. But I could see the benefit of using fallocate holes to optimize the writes of incremental tensors.

Ticketing this as https://github.com/coreweave/tensorizer/issues/159

**Comment by bchess** [comment]
> > ## Accounting for File Range Insertion
> 
> ~Imo its a fair bit of added code complexity, adds more constraints to what we can optimize, for little benefit. Object stores upload APIs have minimums to the chunk sizes, and you still need to assign the correct index of the block ahead of time -- you can't insert a block later.~
> 
> Thinking about this more, I maybe have presumed different from what you were proposing. I don't think we should add a public interface that allows for the mid-range insertion of a tensor. But I could see the benefit of using fallocate holes to optimize the writes of incremental tensors.

Presuming this was what you intended, ticketed as https://github.com/coreweave/tensorizer/issues/159

File: tensorizer/serialization.py
Comment (Eta0) [reply tensorizer/serialization.py:4315]: `nbytes` can be calculated up-front in the `_WriteSpec` constructor as `self.serialized_length = (tensor.element_size() * tensor.nelement()) & ~-(tensor.device.type == "meta")` (at least, in the absence of compression, which would complicate more things than this), so this shouldn't need to wait for anything to get that datapoint. Adding synchronization barriers introduces issues with parallel usage of system resources (though you recently enough read my whole spiel on that!), so it's best to avoid them if possible.

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:3342]: https://github.com/coreweave/tensorizer/issues/159

File: tensorizer/serialization.py
Comment (bchess) [reply tensorizer/serialization.py:4315]: Good point!

Comment (bchess) [reply tensorizer/serialization.py:4315]: https://github.com/coreweave/tensorizer/issues/160

**Comment by Eta0** [comment]
> > > ## Accounting for File Range Insertion
> > 
> > 
> > ~Imo its a fair bit of added code complexity, adds more constraints to what we can optimize, for little benefit. Object stores upload APIs have minimums to the chunk sizes, and you still need to assign the correct index of the block ahead of time -- you can't insert a block later.~
> > Thinking about this more, I maybe have presumed different from what you were proposing. I don't think we should add a public interface that allows for the mid-range insertion of a tensor. But I could see the benefit of using fallocate holes to optimize the writes of incremental tensors.
> 
> Presuming this was what you intended, ticketed as #159

Yes—no, we don't need an interface that controls the actual tensor insertion point in the file—it would rather be useful for expanding the metadata section at the top of the file instead of going through hoops to determine its size or requiring people to specify `max_tensors` in advance. So I'm referring to setting up the entire metadata section as a chunk of indeterminate size, for object storage purposes.

Though, as a sidenote, you can add new chunks into the middle of a multi-part object storage upload, because part indices don't need to be consecutive (for anything compatible with [AWS's object storage](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)):

> ### Parts upload
> When uploading a part, in addition to the upload ID, you must specify a part number. You can choose any part number between 1 and 10,000. A part number uniquely identifies a part and its position in the object you are uploading. The part number that you choose doesn’t need to be in a consecutive sequence (for example, it can be 1, 5, and 14). If you upload a new part using the same part number as a previously uploaded part, the previously uploaded part is overwritten.

If we wanted, we could make the file header (start at) chunk 1 and the data section start at chunk 100, and have room to optionally insert up to 99 extension chunks before committing an upload. But we wouldn't need to.

**Pull request merged**
